### ISSUE TITLE
Intermittent USB connection problems

### ISSUE THREAD

**Issue by scottgonzalez** [tessel/t2-cli#387]
More often than not, `t2` fails to find my Tessel when connected via USB. I always wait for the light on the ethernet port to turn on before running any commands.

OS X 10.10.3
node 0.12.3
npm 2.9.1
t2-cli bc9c59f16359d68b1d73fd9fbfd13094683b9462 (master from 2015-10-17)

I turned on debug messages for the usb module and got the following output:

```
t2-cli (master *) $ t2 list
INFO Searching for nearby Tessels...
[timestamp] [threadID] facility level [function call] <message>
--------------------------------------------------------------------------------
[ 0.252926] [00000713] libusb: debug [libusb_get_device_list] 
[ 0.253046] [00000713] libusb: debug [libusb_get_device_descriptor] 
[ 0.253114] [00000713] libusb: debug [libusb_get_device_descriptor] 
[ 0.253135] [00000713] libusb: debug [libusb_get_device_descriptor] 
[ 0.253716] [00000713] libusb: debug [libusb_open] open 20.5
[ 0.253773] [00000713] libusb: warning [darwin_open] USBDeviceOpen: another process has device opened for exclusive access
[ 0.253817] [00000713] libusb: debug [usbi_add_pollfd] add fd 22 events 1
[ 0.253823] [00000713] libusb: debug [darwin_open] device open for access
[ 0.253846] [00003d03] libusb: debug [handle_events] poll() returned 1
[ 0.253864] [00003d03] libusb: debug [handle_events] caught a fish on the control pipe
[ 0.253880] [00003d03] libusb: debug [libusb_get_next_timeout] no URBs, no timeout!
[ 0.253885] [00003d03] libusb: debug [libusb_try_lock_events] someone else is modifying poll fds
[ 0.253888] [00003d03] libusb: debug [libusb_event_handler_active] someone else is modifying poll fds
[ 0.253891] [00003d03] libusb: debug [libusb_handle_events_timeout_completed] another thread is doing event handling
[ 0.253922] [00003d03] libusb: debug [libusb_get_next_timeout] no URBs, no timeout!
[ 0.253930] [00003d03] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.253934] [00003d03] libusb: debug [handle_events] poll fds modified, reallocating
[ 0.253938] [00003d03] libusb: debug [handle_events] poll() 3 fds with timeout in 60000ms
[ 0.254601] [00000713] libusb: debug [libusb_claim_interface] interface 0
[ 0.254963] [00000713] libusb: debug [get_endpoints] building table of endpoints.
[ 0.254990] [00000713] libusb: debug [darwin_claim_interface] interface opened
[ 0.255139] [00004303] libusb: debug [libusb_set_interface_alt_setting] interface 0 altsetting 2
[ 0.255799] [00004303] libusb: debug [get_endpoints] building table of endpoints.
[ 0.255816] [00004303] libusb: debug [get_endpoints] interface: 0 pipe 1: dir: 1 number: 1
[ 0.255822] [00004303] libusb: debug [get_endpoints] interface: 0 pipe 2: dir: 0 number: 2
[ 0.260658] [0000230f] libusb: debug [darwin_async_io_callback] an async io operation has completed
[ 0.260695] [00003d03] libusb: debug [handle_events] poll() returned 1
[ 0.260708] [00003d03] libusb: debug [op_handle_events] checking fd 22 with revents = 1
[ 0.260714] [00003d03] libusb: debug [darwin_handle_callback] handling control completion with kernel status 0
[ 0.260719] [00003d03] libusb: debug [usbi_handle_transfer_completion] transfer 0x10120cdb8 has callback 0x102eee1cf
[ 0.260742] [00003d03] libusb: debug [libusb_get_next_timeout] no URBs, no timeout!
[ 0.260747] [00003d03] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.260751] [00003d03] libusb: debug [handle_events] poll() 3 fds with timeout in 60000ms
[ 0.261826] [00000713] libusb: debug [ep_to_pipeRef] converting ep address 0x81 to pipeRef and interface
[ 0.261845] [00000713] libusb: debug [ep_to_pipeRef] pipe 1 on interface 0 matches
[ 0.261948] [00000713] libusb: debug [ep_to_pipeRef] converting ep address 0x81 to pipeRef and interface
[ 0.261960] [00000713] libusb: debug [ep_to_pipeRef] pipe 1 on interface 0 matches
[ 0.265026] [00000713] libusb: debug [ep_to_pipeRef] converting ep address 0x02 to pipeRef and interface
[ 0.265042] [00000713] libusb: debug [ep_to_pipeRef] pipe 2 on interface 0 matches
[ 0.265163] [0000230f] libusb: debug [darwin_async_io_callback] an async io operation has completed
[ 0.265194] [00003d03] libusb: debug [handle_events] poll() returned 1
[ 0.265202] [00003d03] libusb: debug [op_handle_events] checking fd 22 with revents = 1
[ 0.265206] [00003d03] libusb: debug [darwin_handle_callback] handling bulk completion with kernel status 0
[ 0.265210] [00003d03] libusb: debug [usbi_handle_transfer_completion] transfer 0x102820058 has callback 0x102eee1cf
[ 0.265218] [00003d03] libusb: debug [libusb_get_next_timeout] no URB with timeout or all handled by OS; no timeout!
[ 0.265223] [00003d03] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.265226] [00003d03] libusb: debug [handle_events] poll() 3 fds with timeout in 60000ms
[ 0.266482] [00000713] libusb: debug [ep_to_pipeRef] converting ep address 0x02 to pipeRef and interface
[ 0.266493] [00000713] libusb: debug [ep_to_pipeRef] pipe 2 on interface 0 matches
[ 0.266590] [0000230f] libusb: debug [darwin_async_io_callback] an async io operation has completed
[ 0.266616] [00003d03] libusb: debug [handle_events] poll() returned 1
[ 0.266630] [00003d03] libusb: debug [op_handle_events] checking fd 22 with revents = 1
[ 0.266634] [00003d03] libusb: debug [darwin_handle_callback] handling bulk completion with kernel status 0
[ 0.266637] [00003d03] libusb: debug [usbi_handle_transfer_completion] transfer 0x1029001c8 has callback 0x102eee1cf
[ 0.266643] [00003d03] libusb: debug [libusb_get_next_timeout] no URB with timeout or all handled by OS; no timeout!
[ 0.266647] [00003d03] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.266650] [00003d03] libusb: debug [handle_events] poll() 3 fds with timeout in 60000ms
[ 0.266712] [00000713] libusb: debug [ep_to_pipeRef] converting ep address 0x02 to pipeRef and interface
[ 0.266720] [00000713] libusb: debug [ep_to_pipeRef] pipe 2 on interface 0 matches
[ 0.270230] [0000230f] libusb: debug [darwin_async_io_callback] an async io operation has completed
[ 0.270273] [00003d03] libusb: debug [handle_events] poll() returned 1
[ 0.270279] [00003d03] libusb: debug [op_handle_events] checking fd 22 with revents = 1
[ 0.270283] [00003d03] libusb: debug [darwin_handle_callback] handling bulk completion with kernel status -536870163
[ 0.270287] [00003d03] libusb: warning [darwin_transfer_status] transfer error: device not responding (value = 0xe00002ed)
[ 0.270290] [00003d03] libusb: debug [usbi_handle_transfer_completion] transfer 0x10120cea8 has callback 0x102eee1cf
[ 0.270299] [00003d03] libusb: debug [libusb_get_next_timeout] no URB with timeout or all handled by OS; no timeout!
[ 0.270302] [00003d03] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 0.270305] [00003d03] libusb: debug [handle_events] poll() 3 fds with timeout in 60000ms
Error: LIBUSB_TRANSFER_ERROR
    at Error (native)
[ 5.256104] [00003d03] libusb: debug [handle_events] poll() returned 1
[ 5.256135] [00003d03] libusb: debug [handle_events] caught a fish on the hotplug pipe
[ 5.256149] [00003d03] libusb: debug [handle_events] poll() 3 fds with timeout in 0ms
[ 5.256166] [00003d03] libusb: debug [handle_events] poll() returned 0
[ 5.256182] [00003d03] libusb: debug [libusb_get_next_timeout] no URB with timeout or all handled by OS; no timeout!
[ 5.256190] [00003d03] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 5.256197] [00003d03] libusb: debug [handle_events] poll() 3 fds with timeout in 60000ms
No Tessels Found. undefined
WARN No Tessels Found.
```

Here are the devices that my system sees connected via USB:

```
t2-cli (master *) $ system_profiler SPUSBDataType
USB:

    USB 3.0 SuperSpeed Bus:

      Host Controller Location: Built-in USB
      Host Controller Driver: AppleUSBXHCI
      PCI Device ID: 0x9cb1 
      PCI Revision ID: 0x0003 
      PCI Vendor ID: 0x8086 
      Bus Number: 0x0a 

        Internal Memory Card Reader:

          Product ID: 0x8406
          Vendor ID: 0x05ac  (Apple Inc.)
          Version: 8.20
          Serial Number: 000000000820
          Speed: Up to 5 Gb/sec
          Manufacturer: Apple
          Location ID: 0x15300000 / 1
          Current Available (mA): 900
          Current Required (mA): 896
          Built-In: Yes

    USB 3.0 Hi-Speed Bus:

      Host Controller Location: Built-in USB
      Host Controller Driver: AppleUSBXHCI
      PCI Device ID: 0x9cb1 
      PCI Revision ID: 0x0003 
      PCI Vendor ID: 0x8086 
      Bus Number: 0x0a 

        Tessel 2:

          Product ID: 0x7551
          Vendor ID: 0x1209
          Version: 1.10
          Serial Number: 5ABY32CXRWWBHF3HVD8JAPQKDT
          Speed: Up to 12 Mb/sec
          Manufacturer: Technical Machine
          Location ID: 0x14200000 / 5
          Current Available (mA): 500
          Current Required (mA): 500

        Bluetooth USB Host Controller:

          Product ID: 0x8290
          Vendor ID: 0x05ac  (Apple Inc.)
          Version: 0.90
          Speed: Up to 12 Mb/sec
          Manufacturer: Broadcom Corp.
          Location ID: 0x14300000 / 2
          Current Available (mA): 500
          Current Required (mA): 0
          Built-In: Yes
```

**Comment by johnnyman727** [comment]
@scottgonzalez can you also run `t2 version` and let me know which release your running on the board?

**Comment by scottgonzalez** [comment]
0.0.4.

**Comment by johnnyman727** [comment]
@scottgonzalez this line is also curious:
```
No Tessels Found. undefined
```
I wonder what it's trying to print out that's undefined. It doesn't do that on my machine.

**Comment by scottgonzalez** [comment]
Oh, yeah, that won't show up for you. That was the first step of my debugging. It's just a log of the two parameters passed to [`closeFailedCommand()`](https://github.com/tessel/t2-cli/blob/bc9c59f16359d68b1d73fd9fbfd13094683b9462/bin/tessel-2.js#L242). At one point I was actually getting errors (not just "No Tessels Found.") and I was trying to trace it back.

**Comment by scottgonzalez** [comment]
After flashing the latest firmware and pulling the latest updates for t2-cli (master), I just had this happen:

```
servo-pca9685 (master) $ t2 list
INFO Searching for nearby Tessels...
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: LIBUSB_TRANSFER_ERROR
    at Error (native)
servo-pca9685 (master) $ t2 list
INFO Searching for nearby Tessels...
Segmentation fault: 11
servo-pca9685 (master) $ t2 list
INFO Searching for nearby Tessels...
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: LIBUSB_TRANSFER_ERROR
    at Error (native)
servo-pca9685 (master) $ t2 list
INFO Searching for nearby Tessels...
WARN Detected a Tessel that may be booting.
	Tessel-02A35527D097	LAN	
```

The Tessel had been connected for several minutes before running the first command and commands were run immediately after each other, without unplugging the Tessel.

**Comment by johnnyman727** [comment]
All of this should be fixed with #482 and #486.
